### PR TITLE
CSHARP-3032: Store first step when speculatively authenticating

### PR DESCRIPTION
**Pull request by vincentkam**
https://evergreen.mongodb.com/version/dot_net_driver_caa2b604298223bdf521f2cf7fdfcd6cd9ed68a4
This PR fixes a bug in my initial PR https://github.com/vincentkam/mongo-csharp-driver/pull/95  where we were discarding the scram sha authenticator when creating the initial command to speculatively authenticate. We must save this authenticator to preserve the r-value required to continue speculative auth conversation.

Without this fix, the authenticator erroneously throws "invalid nonce errors" when attempting finish the speculative auth conversation because we've generated a new nonce instead of saving the old one.

I've verified that this works as expected locally with a small test program and a mongod with auth enabled.

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs [hunk 1]
```diff
@@ -116,8 +116,8 @@ public BsonDocument CustomizeInitialIsMasterCommand(BsonDocument isMasterCommand
         {
             var saslSupportedMechs = CreateSaslSupportedMechsRequest(_credential.Source, _credential.Username);
             isMasterCommand = isMasterCommand.Merge(saslSupportedMechs);
-            var authenticator = new ScramSha256Authenticator(_credential, _randomStringGenerator);
-            return authenticator.CustomizeInitialIsMasterCommand(isMasterCommand);
+            _speculativeScramSha256Authenticator = new ScramSha256Authenticator(_credential, _randomStringGenerator);
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: We have to store this authenticator so that we don't lose the initial nonce (r-value) that we send when speculatively authenticating with SCRAM-SHA-256.

See SCRAM-SHA-256 conversation: https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#id8

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs [hunk 2]
```diff
@@ -146,5 +146,15 @@ private IAuthenticator CreateAuthenticator(IConnection connection, ConnectionDes
...
         }
+
+        private IAuthenticator GetOrCreateAuthenticator(IConnection connection, ConnectionDescription description)
+        {
+            /* It is possible to have for IsMaster["SpeculativeAuthenticate"] != null and for
+             * _speculativeScramSha256Authenticator to be null in the case of multiple authenticators */
+            return description.IsMasterResult.SpeculativeAuthenticate != null &&
+                   _speculativeScramSha256Authenticator != null
+                ? _speculativeScramSha256Authenticator
+                : CreateAuthenticator(connection, description);
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: We defer to the old `CreateAuthenticator` logic iff we don't need to use the saved speculative ScramSha256 authenticator.

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 3]
```diff
@@ -136,8 +138,8 @@ public BsonDocument CustomizeInitialIsMasterCommand(BsonDocument isMasterCommand
             {
                 case "SCRAM-SHA-1":
                 case "SCRAM-SHA-256":
-                    var firstStep = _mechanism.Initialize(connection: null, conversation: null, description: null);
-                    var firstCommand = CreateStartCommand(firstStep);
+                    _speculativeFirstStep = _mechanism.Initialize(connection: null, conversation: null, description: null);
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs]: Similarly, we have to save this first step because it contains the r-value we're sending to the server, and we need this r-value to transition to the next step.


See SCRAM-SHA-256 conversation: https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#id8
See SCRAM-SHA-1 conversation: https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#id5

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs [hunk 4]
```diff
@@ -36,6 +36,7 @@ public class DefaultAuthenticator : IAuthenticator
         private readonly UsernamePasswordCredential _credential;
         private readonly IRandomStringGenerator _randomStringGenerator;
 
+        private ScramSha256Authenticator _speculativeScramSha256Authenticator;
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Fix whitespace and refactor to:

    private IAuthenticator _speculativeAuthenticator;

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs [hunk 5]
```diff
@@ -146,5 +146,15 @@ private IAuthenticator CreateAuthenticator(IConnection connection, ConnectionDes
                     : new MongoDBCRAuthenticator(_credential);
 #pragma warning restore 618
         }
+
+        private IAuthenticator GetOrCreateAuthenticator(IConnection connection, ConnectionDescription description)
+        {
+            /* It is possible to have for IsMaster["SpeculativeAuthenticate"] != null and for
+             * _speculativeScramSha256Authenticator to be null in the case of multiple authenticators */
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:153]: I don't think this is possible. If we `_speculativeAuthenticator` is null then we didn't sent the speculative authenticate fields and we should not expect a speculative authenticate result.

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 2 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Ternary expressions are evil specially when they don't fit on one line. Refactor for readability:

    var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
    var canUseSpeculativeAuthenticator = _speculativeAuthenticator != null && speculativeAuthenticateResult != null;
    return canUseSpeculativeAuthenticator ? _speculativeAuthenticator : CreateAuthenticator(connection, description);

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 6]
```diff
@@ -71,9 +73,9 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
                 var command = CreateStartCommand(currentStep);
 
                 var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
-                if (speculativeAuthenticateResult != null)
+                if (speculativeAuthenticateResult != null && _speculativeFirstStep != null)
                 {
-                    currentStep = Transition(conversation, currentStep, speculativeAuthenticateResult, out command);
+                    currentStep = Transition(conversation, _speculativeFirstStep, speculativeAuthenticateResult, out command);
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:76]: Note: lines 72-73 are only needed when there is no speculative auth result.

Refactor lines 72-79 to:

    ISaslStep currentStep;
    BsonDocument command;
    var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
    if (_speculativeFirstStep != null && speculativeAuthenticateResult != null)
    {
        currentStep = Transition(conversation, _speculativeFirstStep, speculativeAuthenticateResult, out command);
    }
    else
    {
        currentStep = _mechanism.Initialize(connection, conversation, description);
        command = CreateStartCommand(currentStep);
    }

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 7]
```diff
@@ -106,9 +108,9 @@ public async Task AuthenticateAsync(IConnection connection, ConnectionDescriptio
...
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:115]: Same refactoring as in sync method.

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: How do we know that `ScramSha256Authenticator` is the right one to create?

Doesn't it depend at all on the connection string settings?

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 2 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Question: is it even possible for `_speculativeAuthenticator` to be null?

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Discussed offline:
256 is the default as noted here: https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst#speculative-authentication

And we send the mechanism as part of the `speculativeAuth` document, as noted in the ScramShaX unit tests definition of `saslStartmessage`

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 2 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Discussed offline: yes in the case of multiple credentials with no auth source i.e. a list of "Default Authenticators"

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 5 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:153]: Discussed offline: possible in the case of a list of `DefaultAuthenticators` because of the way "isMaster" is merged in `DefaultAuthenticator.Authenticate`

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 8]
```diff
@@ -136,8 +138,8 @@ public BsonDocument CustomizeInitialIsMasterCommand(BsonDocument isMasterCommand
             {
                 case "SCRAM-SHA-1":
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs]: It seems odd that this class would contain code that applies only to one of the subclasses.  Could it be refactored so that this code lives in ScramShaAuthenticator instead?

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 4 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Oops, I was separating the fields by readonly and non-readonly: fixed the whitespace and took your refactor.

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 2 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs]: Took your refactor!

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 6 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:76]: Took your refactor!

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 7 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:115]: Done

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 8 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs]: Hm that's an interesting question!
Prior to you asking this, `ISaslAuthenticator.CustomizeInitialIsMasterCommand` needed to call `ISaslAuthenticator.CreateStartCommand` if and only if the mechanism is SCRAM-SHA-1 or SCRAM-SHA-256.

I've included a refactor in my last set of commits so we can encapsulate the logic a bit better without exposing a new public API (leveraging C#7.2's [`private protected` modifiers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.2/private-protected)).

**Comment by vincentkam** [comment]
Ready for re-review! The 4.4 auth variants timed out in EG but passed all tests: I've restarted them.

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs [hunk 9]
```diff
@@ -35,7 +35,7 @@ public class DefaultAuthenticator : IAuthenticator
         // fields
         private readonly UsernamePasswordCredential _credential;
         private readonly IRandomStringGenerator _randomStringGenerator;
-
+        private IAuthenticator _speculativeAuthenticator;
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:38]: whitespace still not fixed

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 10]
```diff
@@ -31,7 +31,8 @@ namespace MongoDB.Driver.Core.Authentication
     public abstract class SaslAuthenticator : IAuthenticator
     {
         // fields
-        private readonly ISaslMechanism _mechanism;
+        private protected readonly ISaslMechanism _mechanism;
+        private protected ISaslStep _speculativeFirstStep;
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:35]: I'm not sure private protected is the right thing here, even if it is a shiny new feature.

The class is public and anybody can subclass it (even applications in theory). But I don't think anyone can successfully subclass it if you hide these fields from them.

Perhaps just `protected`?

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 11]
```diff
@@ -130,20 +139,21 @@ public async Task AuthenticateAsync(IConnection connection, ConnectionDescriptio
...
 
         /// <inheritdoc/>
-        public BsonDocument CustomizeInitialIsMasterCommand(BsonDocument isMasterCommand)
+        public virtual BsonDocument CustomizeInitialIsMasterCommand(BsonDocument isMasterCommand)
         {
-            switch (_mechanism.Name)
+            return isMasterCommand;
+        }
+
+        private protected virtual BsonDocument CreateStartCommand(ISaslStep currentStep)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:147]: Why did you move this method?

It just makes doing diffs harder.

**Review by rstam — COMMENTED** [review]
A few questions.

File: src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs (hunk 9 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:38]: Sorry, I thought you meant that I had an extra new line separating readonly and non readonly variables. I now see you also wanted a new line afterwards (I think?). Fixed

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 11 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:147]: I moved this method because a "private protected" methods feels like it shouldn't be with the other private methods. In my head, the  hierarchy for this file should be: 

```
public
protected 
private protected
private
```

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 10 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:35]: Wouldn't `private protected` give us (the driver devs) the ability to subclass as needed while disallowing applications from accessing the field?

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/private-protected

If we made the field simply `protected` then we'd essentially be exposing a new bit of public API, wouldn't we? E.g. Applications would be create a custom class that derives from `SaslAuthenticator` that could read and write from `_mechanism` and `_speculativeFirstStep`.

**Comment by vincentkam** [comment]
Ready for re-review

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 11 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:147]: (Additionally, I had changed the accessibility of this method in order to encapsulate SCRAM related logic into the `ScramShaAuthenticator` class, per @jyemin 's request here: https://github.com/vincentkam/mongo-csharp-driver/pull/104#discussion_r407551526 )

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:147]: Moving code around makes diffs harder. And diffs are a powerful tool to review code changes.

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 10 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:35]: Yes... private protected allows us (in *this* DLL only) to subclass this class and still be able to access these field in the subclasses.

My main point is that we are sending contradictory messages: this class is public with a protected constructor which signals that in theory at least an application could write their own subclass. But then we are sending a contradictory message by declaring the fields private protected.

Given that we don't actually want an application to subclass this class we can go with your changes to the field visibility.

**Review by rstam — APPROVED** [review]
LGTM with comments

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 11 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:147]: Gotcha. In the future, I can keep the method in the same position until the final round to make reviewing easier. Or leave it in the same location as long as we're not changing it from public to private or vice versa I suppose :)

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:35]: I think at least for _mechanism, there is no a lot of sense to make it `private internal`, we pass the mechanism in the protected constructor, so I don't see why it should be hidden. Also, maybe use properties?

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs [hunk 12]
```diff
@@ -67,13 +68,17 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
...
             using (var conversation = new SaslConversation(description.ConnectionId))
             {
-                var currentStep = _mechanism.Initialize( connection, conversation, description);
-                var command = CreateStartCommand(currentStep);
-
+                ISaslStep currentStep;
+                BsonDocument command;
                 var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
-                if (speculativeAuthenticateResult != null)
+                if (_speculativeFirstStep != null && speculativeAuthenticateResult != null)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:74]: is there a chance that only one of these fields will be filled?

**Review by DmitryLukyanov — COMMENTED** [review]
A few questions. Also, I'm concerned about this:
>I've verified that this works as expected locally with a small test program and a mongod with auth enabled.

it sounds like we still don't have a real test for this feature. I would suggest that it's better to have it, otherwise, we still can be in a potential situation similar to the reason of this PR.

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 10 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:35]: I think we typically only use properties when it's part of the public API? Since `_mechanism` and `_speculativeFirstStep` will NOT be exposed to the user, shouldn't using fields should be okay, similar to how we use fields instead of properties when they're private?

Unfortunately, while I would have loved to leave `_mechanism`  private, `ScramShaAuthenticator` needs access to it in the `CustomizeInitialIsMasterCommand` method. Theoretically we could potentially make `_speculativeFirstStep` private by moving it into `ScramShaAuthenticator` but then I think we'd need to refactor the SaslAuthenticator's Authenticate code to be virtual and add overriden logic to `ScramShaAuthenticator`, which I'd rather not do (since then we'd also be allowing the user to override the `Authenticate` logic)

**Comment by vincentkam** [comment]
> it sounds like we still don't have a real test for this feature. I would suggest that it's better to have it, otherwise, we still can be in a potential situation similar to the reason of this PR.

Ah sorry for the confusion: if speculative auth is implemented incorrectly, we will see "invalid server nonce" errors when testing in EG with auth enabled. So we do have an integration test.

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 12 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:74]: Hey @DmitryLukyanov ,
Thanks for the insightful question :) It took me a while to re-derive the answer!
So, we can have two cases where only one of the fields is not null.
The first case is simple:

`speculativeFirstStep != null && speculativeAuthenticateResult == null` : We tried to speculatively authenticate, but the server doesn't support this feature.

The second case is a bit more complicated:

`speculativeFirstStep == null && speculativeAuthenticateResult !=null`: We have at multiple credentials with no auth source (e.g. defaultAuthenticators). The first `DefaultAuthenticator` successfully speculatively authenticates with ScramSha256, and merges the `isMasterResult`: https://github.com/vincentkam/mongo-csharp-driver/blob/0944b1f474b0711d03675bbd9c21879e48bb2ec6/src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs#L102
The second `DefaultAuthenticator` will not attempt to speculatively auth because the new isMasterResult will have `SaslSupportedMechs` (from the first default authenticator's merge). Therefore, a new  `ScramSha256Authenticator` is created. This `ScramSha256Authenticator` will see `speculativeAuthenticateResult != null but its `speculativeFirstStep` will be null.

You can test this out if you'd like by running this test: https://github.com/mongodb/mongo-csharp-driver/blob/3b807e658c1459682ebecd12bf3e4e57884ef741/tests/MongoDB.Driver.Tests/AuthenticationTests.cs#L169
and commenting out the null check on `speculativeFirstStep`.

This case is not covered by the spec because we're on of the few (only?) drivers that still supports multiple credentials. According to @jyemin , even Java dropped support a while ago :)

**Comment by vincentkam** [comment]
Ready for re-review!

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 12 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:74]: thanks for clarification

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

File: src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:35]: I think that it's not the best using of `_mechanism` field, but I can be ok with it, so resolved from my side

**Review by jyemin — APPROVED** [review]
LGTM
